File: assignment01_followup/julia_exercise/scalar_product.jl
# TASK: compute z = a*x+y, where
# a = 3, x = [1,....,1], y = [4,....,4] and N = 20
# and write the result to a file called output.txt

# function to comput the requested product
function compute_scalar_product(a, x, y)
    return a .* x .+ y
end

# function to write the result to a file
function write_to_file(filename, z)
    open(filename, "w") do file
        write(file, join(z, "\n"))
    end
end

# Parameters
a = 3.0
N = 20
x = ones(N)   # Vector x with all ones
y = fill(4.0, N)  # vector y with all elements equal to 4

# compute the scalar product
z = compute_scalar_product(a, x, y)

# write the result to a file
write_to_file("output.txt", z)

println("Result saved in 'output.txt'")

File: assignment01_Docker/assignment01_followup/c_exercise/scalar_product.c
//TASK: compute z = a*x+y, where
// a = 3, x = [1,....,1], y = [4,....,4] and N = 20
// and write the result to a file called output.txt

// author: Giovanni Piccolo

#include <stdio.h>
#define DIMENSION 20

void WriteToFile(double z[DIMENSION]);
void ComputeProduct(double a, double x[], double y[], double z[], int size);

int main()
{   
    //values initialization
    double a = 3.0;
    double x[DIMENSION], y[DIMENSION], z[DIMENSION];

    for (int i = 0; i < DIMENSION; i++)
    {
        x[i] = 1.0;
        y[i] = 4.0;
    }

    //product calculation
    ComputeProduct(a, x, y, z, DIMENSION);

    //write to file
    WriteToFile(z);

    printf("The result has been written to output.txt\n");
    return 0;
}

void ComputeProduct(double a, double x[], double y[], double z[], int size)
{   
    for (int i = 0; i < size; i++)
    {
        z[i] = a * x[i] + y[i];
    }
}

void WriteToFile(double z[DIMENSION])
{
    FILE *file;
    file = fopen("output.txt", "w");
    for (int i = 0; i < DIMENSION; i++)
    {
        fprintf(file, "%f\n", z[i]);
    }
    fclose(file);
}